Summary splits sentences just by dots - but not by :?!  - fix #10167

lerni opened the pull request:
https://github.com/silverstripe/silverstripe-framework/issues/10167

michalkleiner commented on src/ORM/FieldType/DBText.php:
```diff
@@ -130,10 +139,12 @@ public function Summary($maxWords = 50, $add = false)
         $sentences = array_filter(array_map(function ($str) {
             return trim($str);
-        }, preg_split('@(?<=\.)@', $value) ?: []));
+        }, preg_split('@(?<=['. $summarySentenceSeparators .'])@', $value) ?: []));
```
Do we need to `preg_quote` something here? What if devs want to split in `@`, what would it do at the current state?

dhensby replied:
agree there is a need to make sure that the strings are escaped

lerni replied:
In its initial state `preg_split` spits out a "[Warning] Unknown modifier" with characters like `@\\`. If `preg_quote` is used some characters 'll be escaped but just some and php.net even says: Note that preg_quote() is not meant to be applied to the $replacement string(s) of preg_replace() etc.

I'm no regex expert but now have something like bellow working. This way, it's possible to separate with @ or even \ but just single characters can be used as separators which should be OK.
`preg_split('@(?<=[(\.|\;|\:|\?|\!)])@', $value);`

What do you think @dhensby @michalkleiner?

dhensby replied:
> Note that preg_quote() is not meant to be applied to the $replacement string(s) of preg_replace() etc.

That's fine, we aren't using it on the `$replacement` strings because that is the second parameter to `preg_replace`, we're concerned about escaping input in the *pattern*, so `preg_quote` is appropriate.

The correct solution should be:

```php
preg_split('@(?<=['. preg_quote($summarySentenceSeparators, '@') .'])@', $value)
```

michalkleiner commented on src/ORM/FieldType/DBText.php:
```diff
@@ -130,10 +139,17 @@ public function Summary($maxWords = 50, $add = false)
+        $summarySentenceSeparatorsConfig = $this->config()->summarySentenceSeparators;
+        $summarySentenceSeparatorsArr = [];
+        foreach ($summarySentenceSeparatorsConfig as $separator) {
+            array_push($summarySentenceSeparatorsArr, '\\' . $separator[0]);
```
The backslash escaping is only needed for some (special) characters, not for everything, right?
If I wanted to split say on `r` or `x` characters, it would make them `\\r` etc.

dhensby replied:
Good point, this would escape something like a literal `s` to be a `\s` (which has special meaning in regex of whitespace chars), so not only is this not the correct approach it will actually cause unexpected behaviours.

michalkleiner replied:
I think we may add a comment above the config prop for the devs to provide the escaping if required and basically don't try to be smart during the use, because we can't assume whether the dev meant `r` vs `\r` or `.` vs `\.`

dhensby replied:
The feature isn't to allow devs to provide arbitrary regex expressions, but to provide a list of chars to break the sentence on (eg: [.?!]).

I think relying on devs to provide escaped regex is an overly fragile.

If someone wants to split on regex then they should probably implement that themselves or we might be better just providing a split function that takes regex.

michalkleiner replied:
Actually yeah, if we list all of them within the `[]` then it's just a list of characters and even a `.` wouldn't be treated as special.

michalkleiner commented:
This is related and solving that might be a better way how to approach the changes to the Summary method.
https://github.com/silverstripe/silverstripe-framework/issues/8190

lerni commented:
I've updated the PR according to the comments, but now that @michalkleiner linked #8190 this may should be closed/reworked and addressed per IntlBreakIterator?

michalkleiner commented:
@lerni we can still use this as a step forward until someone picks up the intl aspect of it.
Here I think we still need to remove the `\\` escaping of each character.

dhensby commented:
And a test would be good 😇

emteknetnz reviewed (changes requested):
Needs tests and a rebase to resolve merge conflicts

GuySartorelli commented:
Please also remove the changelog entry as this has missed the window for inclusion in the 4.11.0 release.

GuySartorelli commented:
Closed in favour of https://github.com/silverstripe/silverstripe-framework/pull/11100

GuySartorelli commented:
Merged - will be released in CMS 5.2 in April
